at-angular.ts

```diff
@@ -107,8 +151,36 @@ module at {
             if (target.$inject && target.$inject.length > 0) {
                 factory.$inject = target.$inject.slice(0);
             }
-            angular.module(moduleName).factory(className, factory);
+            angular.module(moduleName).factory(className || getFuncName(target), factory);
+        };
+    }
+
+    export function decorator(moduleName: string, targetProvider: string): at.IClassAnnotationDecorator {
+
+        return (targetClass: any): void => {
+
+            angular
+                .module(moduleName)
+                .config([
+                    '$provide',
+                    function($provide: angular.auto.IProvideService): void {
+
+                    $provide.decorator(targetProvider, [
+                       '$delegate',
+                        ($delegate: any): void => {
+
+                            $delegate[targetProvider] = $delegate;
+
+                            angular.extend($delegate, new targetClass());
```

Comment:
But It doesn't respond to the need for the injection of the constructor... :/

I want that : 

```typescript

@decorator('myModule', 'SomeService')
class MyClassDecorator {
     contructor(
          $inject('$delegate') private $delegate:SomeService,
          $inject('$log') private $log
     ){
          
     }

     public someMethod2(){
          this.$log.debug('test');
     }
}
```
I think isn't possible :/